File: virtual_contests/1260/D. A Game with Traps.cpp
#include <bits/stdc++.h>

using namespace std;

#define i64 long long
#define i32 int
#define u32 unsigned int
#define u64 unsigned long long

#define len(x) ((int)x.size())
#define all(a) (a).begin(), (a).end()

const i64 MOD  = 1e9 + 7;
const i64 INF  = 1e9;
const i64 NINF = -(1e9);
const i32 MT   = 0;

const char nl  = '\n';

int segmentUnionLength(const vector<
                          pair <int,int> > &seg)
{
    int n = seg.size();

    // Create a vector to store starting and ending
    // points
    vector <pair <int, bool> > points(n * 2);
    for (int i = 0; i < n; i++)
    {
        points[i*2]     = make_pair(seg[i].first, false);
        points[i*2 + 1] = make_pair(seg[i].second, true);
    }

    // Sorting all points by point value
    sort(points.begin(), points.end());

    int result = 0; // Initialize result

    // To keep track of counts of
    // current open segments
    // (Starting point is processed,
    // but ending point
    // is not)
    int Counter = 0;

    // Traverse through all points
    for (unsigned i=0; i<n*2; i++)
    {
        // If there are open points, then we add the
        // difference between previous and current point.
        // This is interesting as we don't check whether
        // current point is opening or closing,
        if (Counter)
            result += (points[i].first -
                        points[i-1].first);

        // If this is an ending point, reduce, count of
        // open points.
        (points[i].second)? Counter-- : Counter++;
    }
    return result;
}

vector<int> ilist(int n) {
    vector<int> result(n);
    for (int i = 0; i < n; i++) {
        cin >> result[i];
    }
    return result;
}

int binsearch(int low, int high, const function<bool(int)>& condition) {
    int result = -1;
    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (condition(mid)) {
            result = mid;
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }
    return result;
}

void solve() {
    int m, n, k, t;
    cin >> m >> n >> k >> t;

    vector<int> a = ilist(m);
    vector<tuple<int, int, int>> arr;

    for (int i = 0; i < k; i++) {
        int l, r, d;
        cin >> l >> r >> d;
        arr.push_back({l, r + 1, d});
    }

    auto cond = [&](int x) {
        vector<pair<int, int>> new_seg;
        for (const auto& [l, r, d] : arr) {
            if (d > x) {
                new_seg.push_back({l, r});
            }
        }
        int s = segmentUnionLength(new_seg);
        return n + 1 + 2 * s <= t;
    };

    int mx = binsearch(0, 2e5 + 1, cond);

    if (mx == -1) {
        cout << 0 << endl;
        return;
    }

    int res = count_if(a.begin(), a.end(), [mx](int i) { return i >= mx; });
    cout << res << endl;
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);
    int tc;
    if(MT) cin >> tc;
    else tc = 1;
    for (int t = 1; t <= tc; t++) {
        solve();
    }
    cout << endl;
}

File: virtual_contests/1260/D. A Game with Traps.py
import sys, math
from collections import deque, defaultdict, Counter

MOD       = 10**9 + 7
FASTIO    = True
DEBUG     = False
INF       = float('inf')
NINF      = float('-inf')
YES, NO   = "YES", "NO"
MT        = False

#====================Solution====================
def sol():
    m, n, k, t = ivars()
    a = ilist()
    arr = []
    for _ in range(k):
        l, r, d = ivars()
        arr.append((l, r + 1, d))
    def cond(x):
        new = []
        for l, r, d in arr:
            if d > x:
                new.append((l, r))
        s = segmentUnionLength(new)
        return n + 1 + 2 * s <= t
    def binsearch(low, high, condition):
        result = None
        while low <= high:
            mid = (low + high) // 2
            if condition(mid):
                result = mid
                high = mid - 1
            else:
                low = mid + 1
        return result
    mx = binsearch(0, 2*(10**5) + 1, cond)
    if mx == None:
        print(0)
    res = 0
    for i in a:
        if i >= mx:
            res += 1
    print(res)

#================================================
def segmentUnionLength(segments):

    # Size of given segments list
    n = len(segments)
    if n == 0: return 0

    # Initialize empty points container
    points = [None] * (n * 2)

    # Create a vector to store starting
    # and ending points
    for i in range(n):
        points[i * 2] = (segments[i][0], False)
        points[i * 2 + 1] = (segments[i][1], True)

    # Sorting all points by point value
    points = sorted(points, key=lambda x: x[0])

    # Initialize result as 0
    result = 0

    # To keep track of counts of current open segments
    # (Starting point is processed, but ending point
    # is not)
    Counter = 0

    # Traverse through all points
    for i in range(0, n * 2):

        # If there are open points, then we add the
        # difference between previous and current point.
        if (i > 0) & (points[i][0] > points[i - 1][0]) &  (Counter > 0):
            result += (points[i][0] - points[i - 1][0])

        # If this is an ending point, reduce, count of
        # open points.
        if points[i][1]:
            Counter -= 1
        else:
            Counter += 1
    return result
def main():
    for _ in range(ipt() if MT else 1):
        sol()


def debug(*args):
    if DEBUG:
        for arg in args:
            print(arg)

def ASSERT(cond):
    if DEBUG:
        assert(cond)


if FASTIO:
    import os
    from io import BytesIO, IOBase

    _str = str
    str = lambda x=b"": x if type(x) is bytes else _str(x).encode()

    BUFSIZE = 8192


    class FastIO(IOBase):
        newlines = 0

        def __init__(self, file):
            self._file = file
            self._fd = file.fileno()
            self.buffer = BytesIO()
            self.writable = "x" in file.mode or "r" not in file.mode
            self.write = self.buffer.write if self.writable else None

        def read(self):
            while True:
                b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
                if not b:
                    break
                ptr = self.buffer.tell()
                self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
            self.newlines = 0
            return self.buffer.read()

        def readline(self):
            while self.newlines == 0:
                b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
                self.newlines = b.count(b"\n") + (not b)
                ptr = self.buffer.tell()
                self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
            self.newlines -= 1
            return self.buffer.readline()

        def flush(self):
            if self.writable:
                os.write(self._fd, self.buffer.getvalue())
                self.buffer.truncate(0), self.buffer.seek(0)


    class IOWrapper(IOBase):
        def __init__(self, file):
            self.buffer = FastIO(file)
            self.flush = self.buffer.flush
            self.writable = self.buffer.writable
            self.write = lambda s: self.buffer.write(s.encode("ascii"))
            self.read = lambda: self.buffer.read().decode("ascii")
            self.readline = lambda: self.buffer.readline().decode("ascii")


    sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
    input = lambda: sys.stdin.readline().rstrip("\r\n")

if __name__ == "__main__":
    finp        = sys.stdin.readline
    njoin       = lambda x: '\n'.join(map(str, x))
    ljoin       = lambda x: ' '.join(map(str, x))
    sjoin       = lambda x: '\n'.join(x)
    ilist       = lambda: list(map(int, finp().split()))
    slist       = lambda: list(input())
    slists      = lambda: input().split()
    ivars       = lambda: map(int, finp().split())
    ipt         = lambda: int(finp())
    main()

